### ISSUE TITLE
Exclusive subscriptions rejected with QuotaExceeded for no reason?

### ISSUE THREAD

**Issue by tgwzwu** [emqx/emqx#13476]
### What happened?

We have a 3-node EMQX cluster and a bunch of .NET services that on startup generate a fresh Client ID and subscribe to a fixed set of $exclusive topics.
Recently, for no apparent reason, these topic subscriptions started failing with QuotaExceeded.

The EMQX Dashboard shows no subscriptions for those topics.

The same user can subscribe to other (similar) $exclusive topics without any issues, so it is not a permission issue.
Topic subscriptions to those failing topics also fail with the MQTTX client, so it is not a client issue.

To me it looks like EMQX thinks there is a still a subscriber for those topics even though there is not.

### What did you expect to happen?

Subscriptions to $exclusive topics for which no subscriptions exist (according to EMQX dashboard) should succeed and not fail with QuotaExceeded.

### How can we reproduce it (as minimally and precisely as possible)?

At the moment I have no idea how to reproduce it.
All I can think of is that our services were killed and restarted many times without unsubscribing cleanly which potentially triggered the issue.

### EMQX version

<details>

We're using this Docker image: https://github.com/emqx/emqx-docker/blob/02b0801191ba3fafb9f9c845c69d8045d7c85430/5.7/Dockerfile.
We also recently updated, which did not fix the issue.

```console
$ ./bin/emqx_ctl broker
# emqx@emqx54-1:/opt/emqx$ emqx_ctl broker
sysdescr  : EMQX
version   : 5.7.1
datetime  : 2024-07-16T09:04:28.721899941+00:00
uptime    : 3 days, 22 hours, 43 minutes, 44 seconds
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="
[https://www.debian.org/"](https://www.debian.org/%22)
SUPPORT_URL="
[https://www.debian.org/support"](https://www.debian.org/support%22)
BUG_REPORT_URL="
[https://bugs.debian.org/"](https://bugs.debian.org/%22)
```

</details>

### Log files

<details>

```
{
    "time": 1721122814727576,
    "level": "debug",
    "msg": "raw_bin_received",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "bin": "......",
    "peername": "<client-ip>:42194",
    "type": "hex",
    "tag": "MQTT",
    "size": 58,
    "pid": "<0.28826976.0>"
}{
    "time": 1721122814727653,
    "level": "debug",
    "msg": "mqtt_packet_received",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "peername": "<client-ip>:42194",
    "tag": "MQTT",
    "pid": "<0.28826976.0>",
    "packet": "SUBSCRIBE(Q1, R0, D0, PacketId=42919 TopicFilters=[$exclusive/site/sys/0xf8/plc-service/request/F005/#(#{nl => 0,qos => 0,\n                                                     rap => 0,rh => 0})])"
}{
    "time": 1721122814727747,
    "level": "debug",
    "msg": "authorization_module_nomatch",
    "topic": "site/sys/0xf8/plc-service/request/F005/#",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "action": "SUBSCRIBE(Q0)",
    "peername": "<client-ip>:42194",
    "tag": "AUTHZ",
    "pid": "<0.28826976.0>",
    "module": "emqx_authz_client_info"
}{
    "time": 1721122814727882,
    "level": "info",
    "msg": "authorization_permission_allowed",
    "topic": "site/sys/0xf8/plc-service/request/F005/#",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "action": "SUBSCRIBE(Q0)",
    "source": "built_in_database",
    "peername": "<client-ip>:42194",
    "tag": "AUTHZ",
    "pid": "<0.28826976.0>"
}{
    "time": 1721122814731004,
    "level": "warning",
    "msg": "cannot_subscribe_topic_filter",
    "topic": "site/sys/0xf8/plc-service/request/F005/#",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "peername": "<client-ip>:42194",
    "reason": "quota_exceeded",
    "pid": "<0.28826976.0>"
}{
    "time": 1721122814731047,
    "level": "debug",
    "msg": "mqtt_packet_sent",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "peername": "<client-ip>:42194",
    "tag": "MQTT",
    "pid": "<0.28826976.0>",
    "packet": "SUBACK(Q0, R0, D0, PacketId=42919, ReasonCodes=[151])"
}{
    "time": 1721122814731401,
    "level": "debug",
    "msg": "insert_channel_info",
    "clientid": "mqttx_zwu_aecd61cf",
    "username": "<user>",
    "mfa": "{emqx_cm,insert_channel_info,3}",
    "peername": "<client-ip>:42194",
    "pid": "<0.28826976.0>",
    "line": 171
}
```

</details>

**Comment by id** [comment]
Hi @tgwzwu .

Could you check if there are sessions for old clients on the Clients page (Monitoring --> Clients)?

**Comment by tgwzwu** [comment]
I've managed to gather some logs which I've attached to the OP where I've tried to subscribe to one of the affected topics with MQTTX.

@id We have thousands of clients on that cluster. The relevant service has 5 instances, all of those are currently connected (5 Clients) and subscribed to those topics with the exclusive subscription feature disabled, so they're subscribed to the topics without "$exclusive/" prefix. 

There is no overlap of topics between those services. 

If I stop the service that subscribes to `site/sys/0xf8/plc-service/request/F005/#` then both the Client and Subscription for that topic will correctly disappear from the Dashboard. But subscribing to `$exclusive/site/sys/0xf8/plc-service/request/F005/#` with MQTTX still fails with QuotaExceeded.

**Comment by id** [comment]
@tgwzwu sessions from abruptly disconnected clients which were using exclusive subscription may still be alive on the server. Do you know which client_id your services with exclusive subscriptions used before? Can you search for those client id in the client list?

**Comment by tgwzwu** [comment]
@id I've searched using both exact and fuzzy search and cannot find any other clients of that service.

**Comment by id** [comment]
@tgwzwu could you run this on emqx node?

```
emqx eval 'ets:tab2list(emqx_exclusive_subscription)'
```

**Comment by tgwzwu** [comment]
@id 
This returns an array of basically all the exclusive subscriptions that don't work!

For example one of them is:
`{exclusive_subscription,<<"site/sys/0xf8/plc-service/request/F005/#">>,
<<"PlcService.039eb34333493c202656d82e7b03af6a.59465cb2-eabb-4015-b5c0-bf83c0b120f6">>},`

Searching for that Client ID in the dashboard (exact query) comes up with no data.

**Comment by id** [comment]
Right, so the theory is correct.

The simplest way to solve this is to disconnect all existing clients with exclusive subscriptions and run this to delete all entries from this table:

```
emqx eval 'emqx_exclusive_subscription:clear()'
```

**Comment by tgwzwu** [comment]
Ok, but why does it keep the exclusive subscriptions around even though all those clients are long gone?

**Comment by id** [comment]
I suppose it's some kind of race condition which is triggered when clients with exclusive subscriptions are killed and restarted many times.

**Comment by tgwzwu** [comment]
Looking through our logs again I can see that all of those subscriptions started failing after a single service instance was killed and restarted. 
From that point forward it was basically stuck in a kill-restart loop because the subscriptions kept failing.
Other services instances joined in on that loop in the following days as they were updated/restarted.

Note that the initial service had just been restarted 10 minutes before all that and exclusive subscriptions were still working then. I also could not find any other services of that type restarting before or during that time window.

But as I mentioned the whole cluster is busy with thousands of other clients, which may have aggravated the race condition.

**Comment by zmstone** [comment]
@tgwzwu thank you.
this is likely a bug.
I cannot seem to find the node-down clean up code for exclusive-subscriptions.

**Comment by zmstone** [comment]
for internal ref, created issue tracker: https://emqx.atlassian.net/browse/EMQX-12707

**Comment by lafirest** [comment]
@tgwzwu @id @zmstone 
This is not a bug, but an original design. 
The lock, I meaning the `exclusive`, request comes from an external resource, and sometimes they may be offline for a while, especially for some devices with limited battery. 
So the thought is that the owner must unlock manually by unsubscribing. 
But in fact, I found that unsubscribing also has a bug when the node is down for whatever reason, we will discuss to find a solution.

**Comment by tgwzwu** [comment]
@lafirest At the very least the dashboard should show (stale) exclusive subscriptions and ideally would also show the client ID that it is/was bound to.

At the moment our application has no persistent state and just generates a random client ID on startup, so if it crashes the subscription would be lost until someone manually removes it. I will need to think of something if this happens more often...

Maybe there should be some expiration date for stale exclusive subscriptions?

**Comment by lafirest** [comment]
@tgwzwu 
Yes, it should be managed in the dashboard, I am going to add CLI and dashboard interface for this.

We cannot assume which is the better way to release the lock,  manual release, or bind to the session lifecycle.

If we release all relevant topics when the node is down, it is possible that a client will reconnect to the cluster and resubscribe to the previous topic, which may have been preempted by other clients.

I'll try adding a new config item to make the behavior configurable.

**Comment by zmstone** [comment]
A recap:
The PRs merged so far are mitigations not fixes.
https://github.com/emqx/emqx/pull/13515 allows the same client to subscribe again.
https://github.com/emqx/emqx/pull/13524 allows user to manually delete a clientID <-> topic bind to allow a new client to subscribe.

There is yet to be a proper fix to perform clean-up by other nodes in the cluster after a node is down, or by the node itself after a single-node restart.

**Comment by lafirest** [comment]
Closed by #13702 #13743
